websites/K/Keep Talking and Nobody Explodes/presence.ts

```diff
@@ -1,159 +1,213 @@
+        return;
+    }
   }
 
   presence.setActivity(presenceData);
+  previous = current;
 });
+
+/**
+ * Get Language Strings
+ * @returns Language Strings
+ */
+async function getStrings() {
+  return presence.getStrings(
+    {
+      reading: "general.reading",
+      page: "general.page"
+    },
+    await presence.getSetting<string>("lang").catch(() => "en")
+  );
+}
+
+/**
+ * Handle bomb manual
+ * @param resetTimestamp Reset the reading timestamp
+ */
+async function handleManual(resetTimestamp = false): Promise<void> {
+  if (timestamp === 0 || resetTimestamp) timestamp = Date.now();
+
+  const pages = [...document.querySelectorAll<HTMLDivElement>(".page")],
+    page = Math.round(
+      (window.scrollY / getDocumentHeight()) * (pages.length - 1)
+    ),
+    presenceData: PresenceData = {
+      details: document
+        .querySelector<HTMLTitleElement>(".title")
+        .textContent.replace(/\n|\t/g, ""),
+      largeImageKey: "logo_big",
+      smallImageText: (await strings).reading,
+      smallImageKey: "reading",
+      startTimestamp: timestamp
+    };
+
+  if (page === 0) {
+    presenceData.state =
+      document.querySelector<HTMLHeadingElement>(
+        ".versioning"
+      ).firstChild.textContent;
+  } else {
+    presenceData.state = `${(await strings).page} ${page + 1} / ${
+      pages.length
+    }: ${pages[page].children[0].children[1].textContent}`;
+  }
+
+  presence.setActivity(presenceData);
+}
+
+/**
+ * Handle faq page
+ */
+async function handleFAQ(): Promise<void> {
+  const presenceData: PresenceData = {
+    details: "FAQ",
+    largeImageKey: "logo_big"
+  };
+
+  if (window.location.hash?.length > 0) {
+    presenceData.state = document.querySelector(
+      window.location.hash
+    ).children[0].children[0].children[0].children[0].children[0].textContent;
+  }
+
+  presence.setActivity(presenceData);
+}
+
+/**
+ * Handle generic pages
+ * @param preferTitle Prefer document title over h1 text content
+ */
+async function handleGeneric(preferTitle = false): Promise<void> {
+  presence.setActivity({
+    details: preferTitle
+      ? document.title
+      : document.querySelector<HTMLHeadingElement>("h1")?.textContent ||
+        document.title,
+    largeImageKey: "logo_big"
+  });
+}
+
+/**
+ * Get document height
+ * @returns Document height
+ */
+function getDocumentHeight(): number {
+  const { body } = document,
+    { documentElement: html } = document;
+
+  return Math.max(
+    body.scrollHeight,
+    body.offsetHeight,
+    html.clientHeight,
+    html.scrollHeight,
+    html.offsetHeight
+  );
+}
+
+/**
+ * Check if a text value is a valid language code
+ * @param text Text value
+ * @returns Valid language code
+ */
+function isLanguageCode(text: string): boolean {
+  return /^(?<lang>[a-z]{2})(-(?<region>[A-Z]{2}))?$/.test(text);
+}
+
+/**
+ * Check if the window location has changed
+ * @param previous Previous window location
+ * @param current Current window location
```

Comment:
Never knew we were composing a library here!!

Reply:
Oh, never knew that it's forbidden to write comments.